JULIA_THEADS=1 build broken by c29bff1

See #15556
Can this be made a milestone 0.5.0 issue?

```
    CC src/threading.o
/home/sachs/src/julia-multithreading/src/threading.c: In function ‘jl_threading_run’:
/home/sachs/src/julia-multithreading/src/threading.c:394:5: error: too many arguments to function ‘jl_get_specialization1’
     jl_lambda_info_t *li = jl_get_specialization1(argtypes, NULL);
     ^
In file included from /home/sachs/src/julia-multithreading/src/threading.c:23:0:
/home/sachs/src/julia-multithreading/src/julia_internal.h:266:19: note: declared here
 jl_lambda_info_t *jl_get_specialization1(jl_tupletype_t *types);
                   ^
make[1]: *** [threading.o] Error 1
make: *** [julia-src-release] Error 2
```

--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -5181,13 +5181,24 @@ static void init_julia_llvm_env(Module *m)
     add_named_global(jltls_states_func, jl_get_ptls_states_getter());
     if (imaging_mode) {
         PointerType *pfunctype = jltls_states_func->getFunctionType()->getPointerTo();
+        // This is **NOT** a external variable or a normal global variable
+        // This is a special internal global slot with a special index
+        // in the global variable table.
         jltls_states_func_ptr =
             new GlobalVariable(*m, pfunctype,
-                               false, GlobalVariable::ExternalLinkage,
-                               NULL, "jl_get_ptls_states.ptr");
+                               false, GlobalVariable::InternalLinkage,
+                               ConstantPointerNull::get(pfunctype),
+                               "jl_get_ptls_states.ptr");
         addComdat(jltls_states_func_ptr);
-        void **p = (void**)jl_emit_and_add_to_shadow(jltls_states_func_ptr);
-        *p = (void*)jl_get_ptls_states_getter();
+        // make the pointer valid for this session
+#if defined(USE_MCJIT) || defined(USE_ORCJIT)
+        auto p = new uintptr_t(0);
+        jl_ExecutionEngine->addGlobalMapping(jltls_states_func_ptr->getName(),
+                                             (uintptr_t)p);
+#else
+        uintptr_t *p = (uintptr_t*)jl_ExecutionEngine->getPointerToGlobal(jltls_states_func_ptr);
+#endif
+        *p = (uintptr_t)jl_get_ptls_states_getter();
         jl_sysimg_gvars.push_back(ConstantExpr::getBitCast(jltls_states_func_ptr,
                                                            T_psize));
         jltls_states_func_idx = jl_sysimg_gvars.size();

--- a/src/threading.c
+++ b/src/threading.c
@@ -391,7 +391,7 @@ JL_DLLEXPORT jl_value_t *jl_threading_run(jl_svec_t *args)
     int8_t gc_state = jl_gc_unsafe_enter();
     JL_GC_PUSH1(&argtypes);
     argtypes = arg_type_tuple(jl_svec_data(args), jl_svec_len(args));
-    jl_lambda_info_t *li = jl_get_specialization1(argtypes, NULL);
+    jl_lambda_info_t *li = jl_get_specialization1(argtypes);
     jl_generate_fptr(li);
 
     threadwork.command = TI_THREADWORK_RUN;

